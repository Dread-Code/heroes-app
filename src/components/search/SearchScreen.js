import React from "react";
import { heroes } from "../../data/heroes";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroList/HeroCard/HeroCard";

export const SearchScreen = () => {
  const heroesFiltered = heroes;

  const [values, handleInputChange, reset] = useForm({
    search: "",
  });

  const handleSearch = (e) => {
    e.preventDefault();
    console.log(values);
    reset();
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              name="search"
              className="form-control"
              onChange={handleInputChange}
              value={values.search}
            />
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          <div
            className="animate__animated animate__fadeIn row row-cols-1 row-cols-md-2 g-4"
            style={{}}
          >
            {heroesFiltered.map((hero) => (
              <HeroCard {...hero} />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};
