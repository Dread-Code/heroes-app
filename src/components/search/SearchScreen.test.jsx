import React from "react";
import { mount } from "enzyme";
import { SearchScreen } from "./SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "@testing-library/react";

describe("SearchScreen", () => {
  const historyMock = {
    push: jest.fn(),
  };

  test("should match snapshoot", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");
  });

  test("should render batman and the value of the queryString input", () => {
    /**
     * Remember that initialEntries receives the arguments that you
     * are using in the query
     */
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("should render a error if doesn't fing a hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batmanadfsade"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    expect(wrapper.find(".alert-danger").exists()).toBe(true);
  });

  test("should call the push history", () => {
    const event = {
      preventDefault: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    // Toca llamar  porque se esta volviendo a renderizar el componente
    // al realizar los siguientes eventos
    // ! todos los eventos que hagan renderizar de nuevo nuestro componente deben ser
    // ! envuelto por act
    act(() => {
      wrapper.find("input").simulate("change", {
        target: {
          name: "search",
          value: "batman",
        },
      });
      wrapper.find("form").prop("onSubmit")(event);
    });

    expect(historyMock.push).toHaveBeenCalledTimes(1);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
