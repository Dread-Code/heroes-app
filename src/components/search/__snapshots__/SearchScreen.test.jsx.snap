// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen should match snapshoot 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
      <div
        className="animate__animated animate__fadeIn row row-cols-1 row-cols-md-2 g-4"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`SearchScreen should render batman and the value of the queryString input 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn row row-cols-1 row-cols-md-2 g-4"
        style={Object {}}
      >
        <div
          className="col"
        >
          <div
            className="card"
            style={
              Object {
                "maxWidth": 540,
              }
            }
          >
            <a
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </a>
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                <strong>
                  Alter ego: 
                </strong>
                 
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <strong
                  style={
                    Object {
                      "fontSize": "2em",
                    }
                  }
                >
                  DC Comics
                </strong>
                <br />
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
